### DISCUSSION TITLE
Would a `proxify` utility ever be useful to valtio itself?

### DISCUSSION THREAD

**Discussion by chiefGui** [pmndrs/valtio#568] [Ideas]
Currently, if you are deserializing data from anywhere, you have to manually turn the deserialized data into `valtio` proxies. That's okay, but as it happened to me in real-world scenario, it didn't take long until it got unmanagable. The solution I came up with is after the process of deserialization, I introduced the "proxification" step, handled by a `proxify` function that looks like the one below:

```ts
import { proxy } from 'valtio'
import { proxyMap, proxySet } from 'valtio/utils'

export function proxify<T extends IDataType>(data: T): T {
  if (data instanceof Array) {
    return proxy(data)
  }

  const keys = Object.keys(data)
  let proxified = data

  for (const k of keys) {
    const val = data[k]

    if (isProxifiable(val)) {
      proxified = { ...(proxified || {}), [k]: proxify(val) }

      continue
    }

    proxified = { ...(proxified || {}), [k]: data[k] }
  }

  if (proxified instanceof Set) {
    return proxySet(proxified) as unknown as T
  }

  if (proxified instanceof Map) {
    return proxyMap(proxified) as unknown as T
  }

  return proxy(proxified) as T
}

function isProxifiable(data: any) {
  if (data instanceof Set) {
    return true
  }

  if (data instanceof Map) {
    return true
  }

  return typeof data === 'object'
}

type IDataType = object | Map<unknown, unknown> | Set<unknown>
```

It not only converts your data piece into a `proxy` (or `proxySet`/`proxyMap`), but also does it deeply/recursively. The above implementation requires some tweaks though, and I'd be glad to tune it up&mdash;even performance-wise&mdash;however I'd like to hear from you if that's something worth including at `valtio`'s package itself.

cc @dai-shi @fkhadra

**Comment by dai-shi** [comment]
> also does it deeply/recursively. 

But, it doesn't do for a map in a map?

> if that's something worth including at valtio's package itself.

At this point, I wouldn't include it in valtio/utils. But, I'm happy to have it in docs as a recipe.

**Reply by chiefGui** [reply]
> But, it doesn't do for a map in a map?

It doesn't. It wasn't my use case so I didn't bother&mdash;the snippet I shared is an exact copy-and-paste of the code I wrote to my reality. As I mentioned, it has to be tweaked.

> At this point, I wouldn't include it in valtio/utils. But, I'm happy to have it in docs as a recipe.

Gotcha. When you say "at this point", it implied to me there's a condition in which it could be included. I wonder, what condition would it be? 🤔

**Reply by dai-shi** [reply]
The condition would be a) clean&maintainable implementation, b) many people want it, and c) reasonable docs to educate the usage. On second thought, it's unlikely because I hesitate to add multiple usages especially in this lib.
